### ISSUE TITLE
syntax error: failed to expand 'HASH_TOKEN_IN_MACRO'

### ISSUE THREAD

**Issue by chrchr-github** [danmar/simplecpp#328]
From https://trac.cppcheck.net/ticket/12105
~~~
void g(const char*);
#define HASH_TOKEN_IN_MACRO( A ) g( #A )
void f() {
	HASH_TOKEN_IN_MACRO (
#ifdef SW
		"Message 1"
#else
		"Message 2"
#endif
	);
}
~~~
`foo.cpp:5: syntax error: failed to expand 'HASH_TOKEN_IN_MACRO', it is invalid to use a preprocessor directive as macro parameter`

**Comment by danmar** [comment]
This is undefined behavior.

https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf

page 152

If there are sequences of preprocessing tokens within the list of arguments that would otherwise act as preprocessing directives,150) the behavior is undefined

**Comment by parrotrueper** [comment]
@danmar Agreed that the code is bad, the linter should flag it and carry on, rather than grind to a halt and stop checking the rest of the code.

**Comment by danmar** [comment]
@parrotrueper  I would hope you can consider to fix the evaluation code somehow then. And evaluate cppcheck using that. A user would fix the code and then recheck..

btw if you evaluate tools you should consider that a tool might detect more problems after code is fixed..

```
void foo(int x, int y) {
    int p;
    if (x > 10)
        int a = p;  // <- uninitialized variable
    if (y > 10)
        int b = p;  // <- uninitialized variable
}
```
Cppcheck only reports 1 warning. If you solve the issue Cppcheck reports about then it reports the other warning. Both bugs are found but not at the same time.

**Comment by danmar** [comment]
alright I guess we can reopen this. Fixing this doesn't mean we have to remove the error message...
